ads/inabox/inabox-host.js

```diff
@@ -23,28 +23,59 @@ import {dev, initLogConstructor, setReportError} from '../../src/log';
 import {reportError} from '../../src/error';
 import {InaboxMessagingHost} from './inabox-messaging-host';
 
+/** @const {string} */
 const TAG = 'inabox-host';
-run(self);
+/** @const {string} */
+const AMP_INABOX_INITIALIZED = 'ampInaboxInitialized';
+/** @const {string} */
+const PENDING_MESSAGES = 'ampInaboxPendingMessages';
+/** @const {string} */
+const INABOX_IFRAMES = 'ampInaboxIframes';
 
 /**
- * @param win {!Window}
+ * Class for initializing host script and consuming queued messages.
+ * @visibleForTesting
  */
-function run(win) {
-  // Prevent double initialization
-  if (win['ampInaboxInitialized']) {
-    dev().info(TAG, 'Skip a 2nd attempt of initializing AMP inabox host.');
-    return;
-  }
-
-  win['ampInaboxInitialized'] = true;
-  initLogConstructor();
-  setReportError(reportError);
+export class InaboxHost {
+  /** @param win {!Window}  */
+  constructor(win) {
+    // Prevent double initialization
+    if (win[AMP_INABOX_INITIALIZED]) {
+      dev().info(TAG, 'Skip a 2nd attempt of initializing AMP inabox host.');
+      return;
+    }
 
-  const host = new InaboxMessagingHost(win, win['ampInaboxIframes']);
+    // Assume we cannot recover from state initialization errors.
+    win[AMP_INABOX_INITIALIZED] = true;
+    initLogConstructor();
+    setReportError(reportError);
 
-  win['ampInaboxPendingMessages'].forEach(message => {
-    host.processMessage(message);
-  });
+    if (win[INABOX_IFRAMES] && !Array.isArray(win[INABOX_IFRAMES])) {
+      dev().info(TAG, `Invalid ${INABOX_IFRAMES}`,
+          win[INABOX_IFRAMES]);
+      win[INABOX_IFRAMES] = [];
+    }
+    const host = new InaboxMessagingHost(win, win[INABOX_IFRAMES]);
 
-  win.addEventListener('message', host.processMessage.bind(host));
+    const queuedMsgs = win[PENDING_MESSAGES];
+    if (queuedMsgs) {
+      if (Array.isArray(queuedMsgs)) {
+        queuedMsgs.forEach(message => {
+          try {
+            host.processMessage(message);
+          } catch (err) {
+            dev().info(TAG, 'Error processing inabox message', message, err);
+          }
+        });
+      } else {
+        dev().info(TAG, `Invalid ${PENDING_MESSAGES}`, queuedMsgs);
+      }
+    }
+    // Empty and ensure that future messages are no longer stored in the array.
+    win[PENDING_MESSAGES] = [];
+    win[PENDING_MESSAGES]['push'] = () => {};
```

Comment:
is this helpful?

Reply:
Ensures that future calls to push have no effect thereby ensuring memory can be freed.

Reply:
makes sense